pkg/build/controller/buildimagechangetrigger/buildimagechangetrigger_controller.go

```diff
@@ -0,0 +1,293 @@
+package buildimagechangetrigger
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	kapi "k8s.io/kubernetes/pkg/api"
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	kcontroller "k8s.io/kubernetes/pkg/controller"
+
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	buildclient "github.com/openshift/origin/pkg/build/client"
+	strategy "github.com/openshift/origin/pkg/build/controller/strategy"
+	buildutil "github.com/openshift/origin/pkg/build/util"
+	osclient "github.com/openshift/origin/pkg/client"
+	oscache "github.com/openshift/origin/pkg/client/cache"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+const (
+	maxRetries = 15
+)
+
+type BuildImageChangeTriggerController struct {
+	bcInstantiator buildclient.BuildConfigInstantiator
+	queue          workqueue.RateLimitingInterface
+
+	//NOTE, currently, neither bc's nor is's pulled from these store are mutated in the handler logic below,
+	//so there is not deep copy logic for bc's
+	bcStore oscache.StoreToBuildConfigListerImpl
+	isStore oscache.StoreToImageStreamLister
+
+	bcStoreSynced func() bool
+	isStoreSynced func() bool
+}
+
+// NewBuildImageChangeTriggerController creates a new BuildImageChangeTriggerController
+func NewBuildImageChangeTriggerController(bcInformer cache.SharedIndexInformer, isInformer cache.SharedIndexInformer, oc osclient.Interface, intkc kclientset.Interface) *BuildImageChangeTriggerController {
+
+	c := &BuildImageChangeTriggerController{
+		bcInstantiator: buildclient.NewOSClientBuildConfigInstantiatorClient(oc),
+		queue:          workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+
+	c.bcStore.Indexer = bcInformer.GetIndexer()
+	c.isStore.Indexer = isInformer.GetIndexer()
+	isInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addImageStream,
+		UpdateFunc: c.updateImageStream,
+	})
+
+	c.bcStoreSynced = bcInformer.HasSynced
+	c.isStoreSynced = isInformer.HasSynced
+
+	return c
+}
+
+func (bc *BuildImageChangeTriggerController) addImageStream(obj interface{}) {
+	stream := obj.(*imageapi.ImageStream)
+	glog.V(4).Infof("Image stream %q added.", stream.Name)
+	bc.enqueueImageStream(stream)
+}
+
+func (bc *BuildImageChangeTriggerController) updateImageStream(old, cur interface{}) {
+	// A periodic relist will send update events for all known streams.
+	newStream := cur.(*imageapi.ImageStream)
+	oldStream := old.(*imageapi.ImageStream)
+	if newStream.ResourceVersion == oldStream.ResourceVersion {
+		return
+	}
```

Comment:
I see this logic was copied from the buildpod controller but i'm not sure i understand it in either.

The point of the relist logic is to give the controller a chance to reconcile/handle any events it may have missed.  By doing this, we effectively prevent that from happening.

Also does a relist result in Update events, or Add events?  I would have expected the latter, but based on this it looks like maybe it's sending an Update event in which both the old+new object are identical?

@csrwng